src/addrman_impl.h

```diff
@@ -0,0 +1,291 @@
+
+#include <cstdint>
+#include <set>
+#include <optional>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+/** Total number of buckets for tried addresses */
+static constexpr int32_t ADDRMAN_TRIED_BUCKET_COUNT_LOG2{8};
+static constexpr int ADDRMAN_TRIED_BUCKET_COUNT{1 << ADDRMAN_TRIED_BUCKET_COUNT_LOG2};
+/** Total number of buckets for new addresses */
+static constexpr int32_t ADDRMAN_NEW_BUCKET_COUNT_LOG2{10};
+static constexpr int ADDRMAN_NEW_BUCKET_COUNT{1 << ADDRMAN_NEW_BUCKET_COUNT_LOG2};
+/** Maximum allowed number of entries in buckets for new and tried addresses */
+static constexpr int32_t ADDRMAN_BUCKET_SIZE_LOG2{6};
+static constexpr int ADDRMAN_BUCKET_SIZE{1 << ADDRMAN_BUCKET_SIZE_LOG2};
+
+/**
+ * Extended statistics about a CAddress
+ */
+class AddrInfo : public CAddress
+{
+public:
+    //! last try whatsoever by us (memory only)
+    int64_t nLastTry{0};
+
+    //! last counted attempt (memory only)
+    int64_t nLastCountAttempt{0};
+
+    //! where knowledge about this address first came from
+    CNetAddr source;
+
+    //! last successful connection by us
+    int64_t nLastSuccess{0};
+
+    //! connection attempts since last successful attempt
+    int nAttempts{0};
+
+    //! reference count in new sets (memory only)
+    int nRefCount{0};
+
+    //! in tried set? (memory only)
+    bool fInTried{false};
+
+    //! position in vRandom
+    mutable int nRandomPos{-1};
+
+    SERIALIZE_METHODS(AddrInfo, obj)
+    {
+        READWRITEAS(CAddress, obj);
+        READWRITE(obj.source, obj.nLastSuccess, obj.nAttempts);
+    }
+
+    AddrInfo(const CAddress& addrIn, const CNetAddr& addrSource) : CAddress(addrIn), source(addrSource)
+    {
+    }
+
+    AddrInfo() : CAddress(), source()
+    {
+    }
+
+    //! Calculate in which "tried" bucket this entry belongs
+    int GetTriedBucket(const uint256& nKey, const std::vector<bool>& asmap) const;
+
+    //! Calculate in which "new" bucket this entry belongs, given a certain source
+    int GetNewBucket(const uint256& nKey, const CNetAddr& src, const std::vector<bool>& asmap) const;
+
+    //! Calculate in which "new" bucket this entry belongs, using its default source
+    int GetNewBucket(const uint256& nKey, const std::vector<bool>& asmap) const
+    {
+        return GetNewBucket(nKey, source, asmap);
+    }
+
+    //! Calculate in which position of a bucket to store this entry.
+    int GetBucketPosition(const uint256& nKey, bool fNew, int nBucket) const;
+
+    //! Determine whether the statistics about this entry are bad enough so that it can just be deleted
+    bool IsTerrible(int64_t nNow = GetAdjustedTime()) const;
+
+    //! Calculate the relative chance this entry should be given when selecting nodes to connect to
+    double GetChance(int64_t nNow = GetAdjustedTime()) const;
+};
+
+class AddrManImpl
+{
+public:
+    AddrManImpl(std::vector<bool>&& asmap, bool deterministic, int32_t consistency_check_ratio)
```

Comment:
What do you think about having the ctor and dtor function bodies in addrman.cpp, so that addrman_impl.h only contains the function declarations?

Reply:
done